app/src/common/shared/org/mozilla/vrbrowser/History.java

```diff
@@ -0,0 +1,210 @@
+package org.mozilla.vrbrowser;
+
+import android.content.Context;
+import android.net.Uri;
+import android.util.Base64;
+import android.util.Log;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.mozilla.geckoview.GeckoSession;
+import org.mozilla.geckoview.GeckoSessionSettings;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+
+public class History implements GeckoSession.ProgressDelegate {
+    private HashMap<String, Item> mItems;
+    private static final int MAX_ITEMS = 30;
+    private static final String LOGTAG = "VRB";
+    private static final String FILENAME = "history.json";
+
+    private Context mContext;
+
+    public static class Item {
+        String url;
+        String title;
+        long timestamp;
+
+        Item() {
+            timestamp = System.currentTimeMillis();
+        }
+
+        Item(JSONObject aObject) {
+            url = aObject.optString("u", "");
+            title = aObject.optString("n", "");
+            timestamp = aObject.optLong("t", System.currentTimeMillis());
+        }
+
+        JSONObject toJSON() {
+            JSONObject obj = new JSONObject();
+            try {
+                obj.put("u", url);
+                obj.put("n", title);
+                obj.put("t", timestamp);
+            }
+            catch (JSONException ex) {
+            }
+            return obj;
+        }
+    }
+
+    public History(Context aContext) {
+        mContext = aContext;
+        mItems = new HashMap<>();
+        SessionStore.get().addProgressListener(this);
+        restore();
+    }
+
+    public void updateContext(Context aContext) {
+        mContext = aContext;
+    }
+
+    public void release() {
+        SessionStore.get().removeProgressListener(this);
+    }
+
+    public String encodedJSON() {
+        String value = getJSONString();
+        String base64 = Base64.encodeToString(value.getBytes(), 0);
+        return Uri.encode(base64);
+    }
+
+    public void clear() {
+        mItems.clear();
+        save();
+    }
+
+    private void save() {
+        JSONArray array = new JSONArray();
+        for (Item item: mItems.values()) {
+            array.put(item.toJSON());
+        }
+
+        try (FileWriter file = new FileWriter(mContext.getFilesDir().getAbsolutePath() + "/" + FILENAME)) {
```

Comment:
as this path is reused below in the `restore` method, can we store it in some private class property?